{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# os.chdir(\"/home/yujia/Project/2023-07-20-individual_MR/dat/02_base_data/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total SNP in the GWAS data: \n",
      "(2459688, 10)\n"
     ]
    }
   ],
   "source": [
    "IGF1 = pd.read_csv(\"/mnt/md0/yujia/project/2023-07-20-individual_MR/dat/02_base_data/IGF1/download/GCST90102623.tsv.gz\", sep=\"\\t\", compression = 'gzip')\n",
    "\n",
    "print(\"Total SNP in the GWAS data: \")\n",
    "print(IGF1.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>base_pair_location</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>beta</th>\n",
       "      <th>standard_error</th>\n",
       "      <th>effect_allele_frequency</th>\n",
       "      <th>p_value</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>92221824</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.026983</td>\n",
       "      <td>0.019768</td>\n",
       "      <td>0.0585</td>\n",
       "      <td>0.184661</td>\n",
       "      <td>rs10</td>\n",
       "      <td>23232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12</td>\n",
       "      <td>125456933</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.011914</td>\n",
       "      <td>0.010451</td>\n",
       "      <td>0.2108</td>\n",
       "      <td>0.268088</td>\n",
       "      <td>rs1000000</td>\n",
       "      <td>27518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>21227772</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.022209</td>\n",
       "      <td>0.008525</td>\n",
       "      <td>0.4968</td>\n",
       "      <td>0.011356</td>\n",
       "      <td>rs10000010</td>\n",
       "      <td>27518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1347325</td>\n",
       "      <td>c</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.011136</td>\n",
       "      <td>0.012498</td>\n",
       "      <td>0.8656</td>\n",
       "      <td>0.386379</td>\n",
       "      <td>rs10000012</td>\n",
       "      <td>27513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>36901464</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>0.004987</td>\n",
       "      <td>0.010135</td>\n",
       "      <td>0.7704</td>\n",
       "      <td>0.632576</td>\n",
       "      <td>rs10000013</td>\n",
       "      <td>27518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2459683</th>\n",
       "      <td>4</td>\n",
       "      <td>4987062</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.019480</td>\n",
       "      <td>0.017285</td>\n",
       "      <td>0.0690</td>\n",
       "      <td>0.273526</td>\n",
       "      <td>rs9999987</td>\n",
       "      <td>26052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2459684</th>\n",
       "      <td>4</td>\n",
       "      <td>123121534</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.010892</td>\n",
       "      <td>0.020474</td>\n",
       "      <td>0.0731</td>\n",
       "      <td>0.605462</td>\n",
       "      <td>rs9999992</td>\n",
       "      <td>17604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2459685</th>\n",
       "      <td>4</td>\n",
       "      <td>98781694</td>\n",
       "      <td>a</td>\n",
       "      <td>t</td>\n",
       "      <td>0.007998</td>\n",
       "      <td>0.008545</td>\n",
       "      <td>0.5339</td>\n",
       "      <td>0.363304</td>\n",
       "      <td>rs9999993</td>\n",
       "      <td>27517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2459686</th>\n",
       "      <td>4</td>\n",
       "      <td>69817056</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>0.025295</td>\n",
       "      <td>0.012085</td>\n",
       "      <td>0.8544</td>\n",
       "      <td>0.042019</td>\n",
       "      <td>rs9999996</td>\n",
       "      <td>27519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2459687</th>\n",
       "      <td>4</td>\n",
       "      <td>164089928</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.005909</td>\n",
       "      <td>0.008552</td>\n",
       "      <td>0.5394</td>\n",
       "      <td>0.501970</td>\n",
       "      <td>rs9999997</td>\n",
       "      <td>27517</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2459688 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chromosome  base_pair_location effect_allele other_allele      beta  \\\n",
       "0                 7            92221824             a            c -0.026983   \n",
       "1                12           125456933             a            g -0.011914   \n",
       "2                 4            21227772             t            c  0.022209   \n",
       "3                 4             1347325             c            g -0.011136   \n",
       "4                 4            36901464             a            c  0.004987   \n",
       "...             ...                 ...           ...          ...       ...   \n",
       "2459683           4             4987062             t            c -0.019480   \n",
       "2459684           4           123121534             a            g -0.010892   \n",
       "2459685           4            98781694             a            t  0.007998   \n",
       "2459686           4            69817056             a            c  0.025295   \n",
       "2459687           4           164089928             a            g -0.005909   \n",
       "\n",
       "         standard_error  effect_allele_frequency   p_value       rs_id      n  \n",
       "0              0.019768                   0.0585  0.184661        rs10  23232  \n",
       "1              0.010451                   0.2108  0.268088   rs1000000  27518  \n",
       "2              0.008525                   0.4968  0.011356  rs10000010  27518  \n",
       "3              0.012498                   0.8656  0.386379  rs10000012  27513  \n",
       "4              0.010135                   0.7704  0.632576  rs10000013  27518  \n",
       "...                 ...                      ...       ...         ...    ...  \n",
       "2459683        0.017285                   0.0690  0.273526   rs9999987  26052  \n",
       "2459684        0.020474                   0.0731  0.605462   rs9999992  17604  \n",
       "2459685        0.008545                   0.5339  0.363304   rs9999993  27517  \n",
       "2459686        0.012085                   0.8544  0.042019   rs9999996  27519  \n",
       "2459687        0.008552                   0.5394  0.501970   rs9999997  27517  \n",
       "\n",
       "[2459688 rows x 10 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IGF1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename the data\n",
    "IGF1.rename(columns={\"chromosome\":\"CHR\", \"base_pair_location\":\"BP\", \"rs_id\":\"SNP\", \"effect_allele\":\"A1\", \"other_allele\":\"A2\", \"effect_allele_frequency\":\"EAF\", \"beta\":\"BETA\", \"standard_error\":\"SE\", \"p_value\":\"P\", \"n\":\"N\"}, inplace=True)\n",
    "\n",
    "# convert EAF to MAF\n",
    "IGF1['MAF'] = IGF1['EAF'].apply(lambda x : x if x <= 0.5 else 1-x)\n",
    "\n",
    "# Uppercase the A1 and A2\n",
    "IGF1['A1'] = IGF1['A1'].str.upper()\n",
    "IGF1['A2'] = IGF1['A2'].str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP left after basic QC: \n",
      "(2459688, 11)\n"
     ]
    }
   ],
   "source": [
    "# reorder the column\n",
    "cols = [\"CHR\", \"BP\", \"SNP\", \"A1\", \"A2\", \"N\", \"SE\", \"P\", \"BETA\", \"MAF\", \"EAF\"]\n",
    "IGF1 = IGF1[cols]\n",
    "\n",
    "# Remove NaN row\n",
    "IGF1 = IGF1.dropna()\n",
    "\n",
    "# convert to correct column type\n",
    "IGF1 = IGF1.astype({'CHR':'int', 'BP':'int', \"SNP\":'str', 'A1':'str', 'A2':'str', 'N':'int', 'SE':'float', 'P':'float', 'BETA':'float', 'MAF':'float', 'EAF':'float'})\n",
    "\n",
    "# remove SNP without SNPID\n",
    "IGF1 = IGF1[IGF1['SNP'].str.startswith('rs')]\n",
    "\n",
    "# save the formatted results\n",
    "IGF1.to_csv(\"/mnt/md0/yujia/project/2023-07-20-individual_MR/dat/02_base_data/IGF1/QC/IGF1_formated_nadrop.tsv.gz\", sep=\"\\t\", index=False)\n",
    "\n",
    "print(\"SNP left after basic QC: \")\n",
    "print(IGF1.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>N</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>BETA</th>\n",
       "      <th>MAF</th>\n",
       "      <th>EAF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>92221824</td>\n",
       "      <td>rs10</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>23232</td>\n",
       "      <td>0.019768</td>\n",
       "      <td>0.184661</td>\n",
       "      <td>-0.026983</td>\n",
       "      <td>0.0585</td>\n",
       "      <td>0.0585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12</td>\n",
       "      <td>125456933</td>\n",
       "      <td>rs1000000</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>27518</td>\n",
       "      <td>0.010451</td>\n",
       "      <td>0.268088</td>\n",
       "      <td>-0.011914</td>\n",
       "      <td>0.2108</td>\n",
       "      <td>0.2108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>21227772</td>\n",
       "      <td>rs10000010</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>27518</td>\n",
       "      <td>0.008525</td>\n",
       "      <td>0.011356</td>\n",
       "      <td>0.022209</td>\n",
       "      <td>0.4968</td>\n",
       "      <td>0.4968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1347325</td>\n",
       "      <td>rs10000012</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>27513</td>\n",
       "      <td>0.012498</td>\n",
       "      <td>0.386379</td>\n",
       "      <td>-0.011136</td>\n",
       "      <td>0.1344</td>\n",
       "      <td>0.8656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>36901464</td>\n",
       "      <td>rs10000013</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>27518</td>\n",
       "      <td>0.010135</td>\n",
       "      <td>0.632576</td>\n",
       "      <td>0.004987</td>\n",
       "      <td>0.2296</td>\n",
       "      <td>0.7704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2459683</th>\n",
       "      <td>4</td>\n",
       "      <td>4987062</td>\n",
       "      <td>rs9999987</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>26052</td>\n",
       "      <td>0.017285</td>\n",
       "      <td>0.273526</td>\n",
       "      <td>-0.019480</td>\n",
       "      <td>0.0690</td>\n",
       "      <td>0.0690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2459684</th>\n",
       "      <td>4</td>\n",
       "      <td>123121534</td>\n",
       "      <td>rs9999992</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>17604</td>\n",
       "      <td>0.020474</td>\n",
       "      <td>0.605462</td>\n",
       "      <td>-0.010892</td>\n",
       "      <td>0.0731</td>\n",
       "      <td>0.0731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2459685</th>\n",
       "      <td>4</td>\n",
       "      <td>98781694</td>\n",
       "      <td>rs9999993</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>27517</td>\n",
       "      <td>0.008545</td>\n",
       "      <td>0.363304</td>\n",
       "      <td>0.007998</td>\n",
       "      <td>0.4661</td>\n",
       "      <td>0.5339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2459686</th>\n",
       "      <td>4</td>\n",
       "      <td>69817056</td>\n",
       "      <td>rs9999996</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>27519</td>\n",
       "      <td>0.012085</td>\n",
       "      <td>0.042019</td>\n",
       "      <td>0.025295</td>\n",
       "      <td>0.1456</td>\n",
       "      <td>0.8544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2459687</th>\n",
       "      <td>4</td>\n",
       "      <td>164089928</td>\n",
       "      <td>rs9999997</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>27517</td>\n",
       "      <td>0.008552</td>\n",
       "      <td>0.501970</td>\n",
       "      <td>-0.005909</td>\n",
       "      <td>0.4606</td>\n",
       "      <td>0.5394</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2459688 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR         BP         SNP A1 A2      N        SE         P  \\\n",
       "0          7   92221824        rs10  A  C  23232  0.019768  0.184661   \n",
       "1         12  125456933   rs1000000  A  G  27518  0.010451  0.268088   \n",
       "2          4   21227772  rs10000010  T  C  27518  0.008525  0.011356   \n",
       "3          4    1347325  rs10000012  C  G  27513  0.012498  0.386379   \n",
       "4          4   36901464  rs10000013  A  C  27518  0.010135  0.632576   \n",
       "...      ...        ...         ... .. ..    ...       ...       ...   \n",
       "2459683    4    4987062   rs9999987  T  C  26052  0.017285  0.273526   \n",
       "2459684    4  123121534   rs9999992  A  G  17604  0.020474  0.605462   \n",
       "2459685    4   98781694   rs9999993  A  T  27517  0.008545  0.363304   \n",
       "2459686    4   69817056   rs9999996  A  C  27519  0.012085  0.042019   \n",
       "2459687    4  164089928   rs9999997  A  G  27517  0.008552  0.501970   \n",
       "\n",
       "             BETA     MAF     EAF  \n",
       "0       -0.026983  0.0585  0.0585  \n",
       "1       -0.011914  0.2108  0.2108  \n",
       "2        0.022209  0.4968  0.4968  \n",
       "3       -0.011136  0.1344  0.8656  \n",
       "4        0.004987  0.2296  0.7704  \n",
       "...           ...     ...     ...  \n",
       "2459683 -0.019480  0.0690  0.0690  \n",
       "2459684 -0.010892  0.0731  0.0731  \n",
       "2459685  0.007998  0.4661  0.5339  \n",
       "2459686  0.025295  0.1456  0.8544  \n",
       "2459687 -0.005909  0.4606  0.5394  \n",
       "\n",
       "[2459688 rows x 11 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IGF1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP left after removing SNPs with MAF < 0.01: \n",
      "2459689\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# =============================\n",
    "# Step 2: remove SNPs with MAF < 0.01\n",
    "# =============================\n",
    "\n",
    "cmd = \"gunzip -c /mnt/md0/yujia/project/2023-07-20-individual_MR/dat/02_base_data/IGF1/QC/IGF1_formated_nadrop.tsv.gz |\" + \\\n",
    "      \"awk 'NR==1 || ($10 > 0.01) {print}' |\" + \\\n",
    "      \"gzip > /mnt/md0/yujia/project/2023-07-20-individual_MR/dat/02_base_data/IGF1/QC/IGF1.standardGWASQC.gz\"\n",
    "os.system(cmd)\n",
    "\n",
    "cmd = \"zcat /mnt/md0/yujia/project/2023-07-20-individual_MR/dat/02_base_data/IGF1/QC/IGF1.standardGWASQC.gz | wc -l\"\n",
    "print(\"SNP left after removing SNPs with MAF < 0.01: \")\n",
    "print(os.popen(cmd).read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP left after removing duplicate SNPs: \n",
      "2459689\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# =============================\n",
    "# Step 3: remove duplicate SNPS\n",
    "# =============================\n",
    "\n",
    "cmd = \"gunzip -c /mnt/md0/yujia/project/2023-07-20-individual_MR/dat/02_base_data/IGF1/QC/IGF1.standardGWASQC.gz |\" + \\\n",
    "      \"awk '{seen[$3]++; if(seen[$3]==1){ print}}' |\" + \\\n",
    "      \"gzip - > /mnt/md0/yujia/project/2023-07-20-individual_MR/dat/02_base_data/IGF1/QC/IGF1.nodup.gz\"\n",
    "os.system(cmd)\n",
    "\n",
    "cmd = \"zcat /mnt/md0/yujia/project/2023-07-20-individual_MR/dat/02_base_data/IGF1/QC/IGF1.nodup.gz | wc -l\"\n",
    "print(\"SNP left after removing duplicate SNPs: \")\n",
    "print(os.popen(cmd).read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP left after removing ambiguous SNPs: \n",
      "2080031\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# =============================\n",
    "# Step 4: keep non-ambiguous SNPS\n",
    "# =============================\n",
    "\n",
    "cmd = \"gunzip -c /mnt/md0/yujia/project/2023-07-20-individual_MR/dat/02_base_data/IGF1/QC/IGF1.nodup.gz | \" + \\\n",
    "      \"awk '!( ($4==\\\"A\\\" && $5==\\\"T\\\") || ($4==\\\"T\\\" && $5==\\\"A\\\") || ($4==\\\"G\\\" && $5==\\\"C\\\") || ($4==\\\"C\\\" && $5==\\\"G\\\")) {print}' | \" + \\\n",
    "      \"gzip > /mnt/md0/yujia/project/2023-07-20-individual_MR/dat/02_base_data/IGF1/QC/IGF1.QC.gz\"\n",
    "os.system(cmd)\n",
    "\n",
    "cmd = \"zcat /mnt/md0/yujia/project/2023-07-20-individual_MR/dat/02_base_data/IGF1/QC/IGF1.QC.gz | wc -l\"\n",
    "print(\"SNP left after removing ambiguous SNPs: \")\n",
    "print(os.popen(cmd).read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mr",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
